entity-service-impl/src/main/java/org/hypertrace/entity/type/service/v2/EntityTypeServiceImpl.java

```diff
@@ -0,0 +1,151 @@
+import org.hypertrace.core.documentstore.Filter.Op;
+import org.hypertrace.core.documentstore.Query;
+import org.hypertrace.core.documentstore.SingleValueKey;
+import org.hypertrace.core.grpcutils.context.RequestContext;
+import org.hypertrace.entity.service.constants.EntityCollectionConstants;
+import org.hypertrace.entity.service.constants.EntityServiceConstants;
+import org.hypertrace.entity.service.util.TenantUtils;
+import org.hypertrace.entity.type.service.v2.EntityTypeServiceGrpc.EntityTypeServiceImplBase;
+import org.hypertrace.entity.type.service.v2.model.EntityTypeDocument;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class EntityTypeServiceImpl extends EntityTypeServiceImplBase {
+  private static final Logger LOGGER = LoggerFactory.getLogger(EntityTypeServiceImpl.class);
+
+  private final Collection entityTypeCollection;
+
+  public EntityTypeServiceImpl(Datastore datastore) {
+    this.entityTypeCollection =
+        datastore.getCollection(EntityCollectionConstants.ENTITY_TYPES_V2_COLLECTION);
+  }
+
+  @Override
+  public void upsertEntityType(org.hypertrace.entity.type.service.v2.UpsertEntityTypeRequest request,
+      io.grpc.stub.StreamObserver<org.hypertrace.entity.type.service.v2.UpsertEntityTypeResponse> responseObserver) {
+    Optional<String> tenantId = RequestContext.CURRENT.get().getTenantId();
+    if (tenantId.isEmpty()) {
+      responseObserver.onError(new ServiceException("Tenant id is missing in the request."));
+      return;
+    }
+
+    // validate the request.
+    if (!isValidUpsert(request.getEntityType())) {
+      responseObserver.onError(new ServiceException("Invalid upsert request."));
+      return;
+    }
+
+    try {
+      // Set tenant id.
+      EntityTypeDocument document = EntityTypeDocument.fromProto(tenantId.get(), request.getEntityType());
+      entityTypeCollection.upsert(new SingleValueKey(tenantId.get(), request.getEntityType().getName()), document);
+
+      // Query the entity type again and return that.
+      Iterator<Document> entityTypes =
+          entityTypeCollection.search(transform(tenantId.get(), List.of(request.getEntityType().getName())));
+
+      if (entityTypes.hasNext()) {
+        String json = entityTypes.next().toJson();
+          responseObserver.onNext(UpsertEntityTypeResponse.newBuilder()
+              .setEntityType(EntityTypeDocument.fromJson(json).toProto()).build());
+        responseObserver.onCompleted();
+      } else {
+        // This could happen if the upsert has failed to make the doc store.
+        // Return an error for now.
+        responseObserver.onError(new RuntimeException("Error upserting EntityType:" + request));
+      }
+    } catch (IOException e) {
+      responseObserver.onError(new RuntimeException("Error upserting EntityType:" + request, e));
+    }
+  }
+
+  private boolean isValidUpsert(EntityType entityType) {
+    return !Strings.isNullOrEmpty(entityType.getName()) &&
```

Comment:
Much of this has not been tested. If it is easy to test can we put in the tests?

Reply:
FTR, discussed offline. Follow up PRs will address the unit tests. This PR already covers all the code with integration tests.